concourse/scripts/entry.sh

```diff
@@ -0,0 +1,163 @@
+        groupadd supergroup
+        case "$test_os" in
+            centos*)
+                /usr/sbin/useradd -G supergroup,tty gpadmin
+                ;;
+            ubuntu*)
+                /usr/sbin/useradd -G supergroup,tty gpadmin -s /bin/bash
+                ;;
+            sles*)
+                # create a default group gpadmin, and add user gpadmin to group gapdmin, supergroup,
+                # tty
+                /usr/sbin/useradd -U -G supergroup,tty gpadmin
+                ;;
+            photon*)
+                /usr/sbin/useradd -U -G supergroup,tty,root gpadmin
+                ;;
+            *) echo "Unknown OS: $test_os"; exit 1 ;;
+        esac
+        echo -e "password\npassword" | passwd gpadmin
+    fi
+    mkdir -p /home/gpadmin
+    chown gpadmin:gpadmin /home/gpadmin
+
+    chown -R gpadmin:gpadmin /tmp/build
+    ln -s "${CONCOURSE_WORK_DIR}"/* /home/gpadmin
+}
+
+# Extract gpdb binary
+function install_gpdb() {
+    [ ! -d /usr/local/greenplum-db-devel ] && mkdir -p /usr/local/greenplum-db-devel
+    tar -xzf "${CONCOURSE_WORK_DIR}"/bin_gpdb/*.tar.gz -C /usr/local/greenplum-db-devel
+    chown -R gpadmin:gpadmin /usr/local/greenplum-db-devel
+}
+
+## Currently, isolation2 testing framework relies on pg_isolation2_regress, we
+## should build it from source. However, in concourse, the gpdb_bin is fetched
+## from remote machine, the $(abs_top_srcdir) variable points to a non-existing
+## location, we fixes this issue by creating a symbolic link for it.
+function create_fake_gpdb_src() {
+    local fake_gpdb_src
+    fake_gpdb_src="$(\
+        grep -rhw '/usr/local/greenplum-db-devel' -e 'abs_top_srcdir = .*' |\
+        head -n 1 | awk '{ print $NF; }')"
+
+    if [ -d "${fake_gpdb_src}" ]; then
+        echo "Fake gpdb source directory has been configured."
+        return
+    fi
+
+    pushd /home/gpadmin/gpdb_src
+    ./configure --prefix=/usr/local/greenplum-db-devel \
+        --without-zstd \
+        --disable-orca --disable-gpcloud --enable-debug-extensions
+    popd
+
+    local fake_root
+    fake_root=$(dirname "${fake_gpdb_src}")
+    mkdir -p "${fake_root}"
+    ln -s /home/gpadmin/gpdb_src "${fake_gpdb_src}"
+}
+
+# Setup common environment
+setup_gpadmin
+# install_cmake
+install_gpdb
+
+# Do the special setup with root permission for the each task, then run the real task script with
+# gpadmin. bashrc won't be read by 'su', it needs to be sourced explicitly.
+case "$1" in
+    build)
+        touch /home/gpadmin/.bashrc
+        su gpadmin -c \
+            "source /home/gpadmin/.bashrc &&\
+            /home/gpadmin/ip4r_src/concourse/scripts/build_ip4r.sh"
+        ;;
+    test)
+        # Build task output is ip4r_artifacts, which is different from test taks input
+        # ip4r_bin. Ideally we can use the same name for input and output. But that will cause
+        # compatible issues with 1.x pipeline script.
+        ln -s /home/gpadmin/bin_ip4r /home/gpadmin/ip4r_artifacts
```

Comment:
There were some historical reasons for diskquota to do this. But we can make it simpler and right from the beginning for ip4r.

See https://github.com/pivotal/gp-plpython3-pkg/blob/main/concourse/scripts/entry.sh for a better example. Let me know if you need help.